lib/roby/test/execution_expectations.rb

```diff
@@ -1145,30 +1164,49 @@ def to_s
             end
 
             class Maintain < Expectation
-                def initialize(at_least_during, block, description, backtrace)
+                def initialize(at_least, block, description, backtrace)
                     super(backtrace)
-                    @at_least_during = at_least_during
+
+                    if at_least.kind_of? EventGenerator
+                        @at_least_until = at_least
+                    else
+                        @at_least_during = at_least
+                        @deadline = Time.now + at_least
+                    end
+
                     @description = description || @backtrace[0].to_s
                     @block = block
-                    @deadline = Time.now + at_least_during
                     @failed = false
                 end
 
+                def emitted?(propagation_info)
+                    return false unless @at_least_until
+
+                    @emitted_events =
+                        propagation_info
+                        .emitted_events
+                        .find_all { |ev| ev.generator == @at_least_until }
+                    !@emitted_events.empty?
+                end
+
                 def update_match(propagation_info)
                     if !@block.call(propagation_info)
                         @failed = true
                         false
-                    elsif Time.now > @deadline
-                        true
+                    else
+                        emitted?(propagation_info) ||
+                            (@at_least_during && Time.now > @deadline)
```

Comment:
Doesn't this never match if neither `@at_least*` are set?

I would check on `@deadline` here since the condition is using it.

Reply:
> Doesn't this never match if neither @at_least* are set?

The user level API enforce one of them to be set (just as it did before this patch)

Reply:
OK. Please add a comment explaining that, it will make things easier on the understanding.